### PR TITLE
Implement Publish/Subscribe

### PR DESCRIPTION
**Pull request by allenlee820202**
1. Implement Publish/Subscribe on broker server
2. Modify Mqtt codec to support groupId in SUBSCRIBE

#22 #24

### CODE REVIEW

**Comment by allenlee820202** [comment]
@tz70s This one should be better!

**Comment by tz70s** [comment]
Checkout some reviewed comments in #25, thanks! I'll carefully review this tmr.

File: broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MqttMessageHandler.java [hunk 1]
```diff
@@ -25,11 +37,21 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
...
+                    break;
+                case SUBSCRIBE:
+                    protocolProcessor.processSubscribe(ctx.channel(), (MqttSubscribeMessage) mqttMessage);
+                    logger.debug("[MqttMessageHandler] SUBSCRIBE");
                     break;
             }
         } catch (NullPointerException e) {
-            logger.error(e.getMessage());
+            logger.info(e.getMessage());
+            e.printStackTrace();
```
Comment (tz70s) [inline broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MqttMessageHandler.java]: Remove print stack trace and move info to error

File: broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessagePublisher.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    private void publishToSubscriptions(ServerContext serverContext,
+                                      MqttPublishMessage mqttPublishMessage) {
+        // TODO: publish messages to the subscriptions in every server session
+        // Iterate through all sessions, publishing to every sessions' subscriptions
+        for (ServerSession serverSession: serverContext.getServerSessionPool().asCollection()) {
+            // Only publish to active sessions
+            if(serverSession.getIsActive()) {
+                // Whether the mqttPublish Message matches the subscription is performed in "sendToSubscribers"
+                serverSession.getSubscriptionPool().sendToSubscribers(mqttPublishMessage);
+            }
```
Comment (tz70s) [inline broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessagePublisher.java]: Async

File: broke-broker/src/main/java/org/dsngroup/broke/broker/storage/Subscription.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Overrride equals() for Set comparator
+     * Compare using the reference of subscriberChannelHandlerContext
+     * @param obj Other object
+     * */
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
```
Comment (tz70s) [inline broke-broker/src/main/java/org/dsngroup/broke/broker/storage/Subscription.java]: Use instanceof may be more clear

File: broke-broker/src/main/java/org/dsngroup/broke/broker/storage/SubscriptionPool.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    /**
+     * The subscriber pool
+     * The List of subscribers
+     * */
+    private List<Subscription> subscriptionList;
+
+    /**
+     *
+     * */
+    private Random packetIdRandomGenerator = new Random();
```
Comment (tz70s) [inline broke-broker/src/main/java/org/dsngroup/broke/broker/storage/SubscriptionPool.java]: Use UUID

Comment (allenlee820202) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/storage/SubscriptionPool.java]: Since the packetId needs to be integer, a 32-byte UUID may not be a good choice. In [this](https://stackoverflow.com/questions/5563502/uuid-to-unique-integer-id) website, the static AtomicInteger is mentioned to be able to generate random integer at the runtime.

Comment (tz70s) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/storage/SubscriptionPool.java]: Ok, just ensure there’s no conflict occurred.

File: broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessagePublisher.java (hunk 2 above)
Comment (allenlee820202) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessagePublisher.java]: I'm considering making this asynchronous (or multi-threaded) will conflict with the EventLoopGroup design of Netty. 

For each channel, an event loop is assigned to run all handlers on such channel, which will be run by a single thread. So the channels' tasks won't block each other unless the threads assigned to the event loop group is less than the number of channels. 

Making this asynchronous means that more than one threads are added to a **single** channel, which may make the context switch more often if there are multiple channels. 

In summary, multiple channel handlers won't be run by different threads since they are in a single channel. Only different channels will be run by different threads.

Comment (tz70s) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessagePublisher.java]: No, actually we don’t need to add additional threads to deal with this. IMHO, the write back to subscriber may need to register an additional handler as outbound in pipeline. The chosen time and scheduling will be assigned by netty. It’s not clear I mark async here, instead, the place of write subscriber’s channel. You should figure out the correct non-blocking read/write in netty.

Comment (allenlee820202) [reply broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessagePublisher.java]: OK, I see!

**Comment by allenlee820202** [comment]
@tz70s I added an inbound handler "PublishToSubscriberHandler" in each server channel for asynchronous publishing to subscribers on the broker servers.

**Comment by tz70s** [comment]
After brief view, I think the implementation is not proper. Have you stressed-test it? Or observed the detailed thread creations?
Anyway, to avoid blocking progress, I'll merge this first. 
The next steps are, 
1. rtt mechanism implementation.
2. scratch a precise workflow of connection handling in slide, including thread creation.
3. add a simple benching program, which creates over 100 clients.
4. then, we'll re-discuss the asynchrouny.
